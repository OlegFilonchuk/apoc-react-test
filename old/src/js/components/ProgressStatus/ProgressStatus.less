.progress-status {
  display: block;

  &__stages-container {
    display: flex;
    list-style: none;
    margin: 0;
    padding: 0;
    padding-bottom: 2em;
  }

  &__label {
    display: block;
    text-align: left;
  }

  &__progress {
    text-align: right;
  }
}
