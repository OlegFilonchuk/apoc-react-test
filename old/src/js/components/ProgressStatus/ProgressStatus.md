## Staged progress component
Combines a [Progressbar](#progressbar) and a [ProgressStatusStage](#progressstatusstage) component to display the progress of a long running staged operation.

## ProgressStatus Example

    (
        <ProgressStatus progress={43} progressLabel="my_file.mls">
            <ProgressStatusStage title="Importing File"></ProgressStatusStage>
            <ProgressStatusStage title="Converting To Scenario"></ProgressStatusStage>
        </ProgressStatus>
    )

## ProgressStatus Example with 3 items

    (
        <ProgressStatus progress={65} progressLabel="my_file.mls">
            <ProgressStatusStage title="Importing File"></ProgressStatusStage>
            <ProgressStatusStage title="Converting To Scenario"></ProgressStatusStage>
            <ProgressStatusStage title="Reloading the view"></ProgressStatusStage>
        </ProgressStatus>
    )

..or more

    const changePercentage = () => {
      setState({ progress: Math.floor(Math.random()* 100) });
    };

    (   
        <div>
            <button onClick={() => setState({progress: 0})}>0</button>
            <button onClick={() => setState({progress: 60})}>60</button>
            <button onClick={() => setState({progress: 99})}>99</button>
            <button onClick={() => setState({progress: 100})}>100</button>
            <button onClick={changePercentage}>Random</button>
            <ProgressStatus progress={state.progress} progressLabel="my_file.mls">
                <ProgressStatusStage title="Importing File"></ProgressStatusStage>
                <ProgressStatusStage title="Converting To Scenario"></ProgressStatusStage>
                <ProgressStatusStage title="Loading the Scenario"></ProgressStatusStage>
                <ProgressStatusStage title="Cleanup"></ProgressStatusStage>
                <ProgressStatusStage title="Reloading the view"></ProgressStatusStage>
            </ProgressStatus>
        </div>
    )