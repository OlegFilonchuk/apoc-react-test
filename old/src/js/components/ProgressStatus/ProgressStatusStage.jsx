import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './ProgressStatusStage.less';

export const STAGE_STATUS = {
  WAITING: 'waiting',
  INPROGRESS: 'inprogress',
  FINISHED: 'finished'
};

Object.freeze(STAGE_STATUS);

const loadingStatusIcon = (status, baseClass) => {
  if (status === STAGE_STATUS.INPROGRESS) {
    return <div className={`${baseClass}__progress-animation`} />;
  } else if (status === STAGE_STATUS.FINISHED) {
    return <span className="sc-icon-check" />;
  }

  return null;
};

function ProgressStatusStage(props) {
  const { className, title, status } = props;

  const baseClass = 'progress-status-stage';
  const containerBaseClass = `${baseClass}__container`;

  const containerClasses = classNames(containerBaseClass, className, {
    [`${containerBaseClass}--finished`]: status === STAGE_STATUS.FINISHED,
    [`${containerBaseClass}--inprogress`]: status === STAGE_STATUS.INPROGRESS
  });

  return (
    <li className={`${baseClass}__wrapper`}>
      <div className={containerClasses}>
        <span className={`${containerBaseClass}-title`}>{title}</span>
        {loadingStatusIcon(status, baseClass)}
      </div>
    </li>
  );
}

ProgressStatusStage.propTypes = {
  /**
   * List of custom classes added to component
   */
  className: PropTypes.string,
  /**
   * The status of the stage
   */
  status: PropTypes.oneOf([STAGE_STATUS.WAITING, STAGE_STATUS.INPROGRESS, STAGE_STATUS.FINISHED]),
  /**
   * The title of the stage
   * ..maybe use status prop oneOf['finished', 'inprogress'] ?!
   */
  title: PropTypes.string.isRequired
};

ProgressStatusStage.defaultProps = {
  className: '',
  status: STAGE_STATUS.WAITING
};

export default ProgressStatusStage;
