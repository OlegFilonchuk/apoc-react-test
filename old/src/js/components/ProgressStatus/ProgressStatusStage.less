@import '~chameleon/less/variables.less';

@stageColor: darken(@primary, 10%);
@stageInactiveColor: fadeout(greyscale(@stageColor), 50%);

.progress-status-stage {
  &__wrapper {
    flex: 1;
    counter-increment: stage-counter;
  }

  &__container {
    width: 7em;
    height: 7em;
    margin: 0 auto;
    padding: 0.25em 0.5em;
    border: 2px solid @stageInactiveColor;
    border-radius: 100%;
    display: flex;
    flex-direction: column;
    font-weight: bold;
    text-align: center;
    color: @stageInactiveColor;

    .sc-icon-check {
      opacity: 0;
      flex: 1;
    }

    &-title {
      display: block;
      flex: 2;
      overflow: hidden;

      &::before {
        content: counter(stage-counter);
        display: block;
        flex: 1;
      }
    }

    &--inprogress,
    &--finished {
      border: 2px solid @stageColor;
      color: @stageColor;
    }

    &--finished .sc-icon-check {
      opacity: 1;
    }
  }
}
