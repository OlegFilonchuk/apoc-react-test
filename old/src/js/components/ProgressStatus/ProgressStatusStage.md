### ProgressStatusStage component for use with [ProgressStatus](#progressstatus).
