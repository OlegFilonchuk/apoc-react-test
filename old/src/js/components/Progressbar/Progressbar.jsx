import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import numberInRange from '../../utils/validation/numberInRange';

import './Progressbar.less';

function Progressbar(props) {
  const { titleSettings, color, className, percentage } = props;

  const findProgressRange = range => (percentage >= range.min && percentage <= range.max ? range : null);

  const getBarDynamicClass = () => {
    if (color && color.dynamic) {
      const percentageRange = color.range.find(findProgressRange);

      return percentageRange ? percentageRange.className : '';
    }

    return '';
  };

  const getBarClass = () => {
    const rangeDynamicClass = getBarDynamicClass();

    return classNames(className, rangeDynamicClass, 'sc-progress-bar', {
      'progress-inner-text': titleSettings && titleSettings.inside
    });
  };

  const isTitleOutside = titleSettings && titleSettings.outside;
  const isInnerPercentage = titleSettings && titleSettings.inside;
  const title = isTitleOutside ? <span className="sa-progress-title-sync">{percentage}%</span> : null;
  const innerLabel = isInnerPercentage ? `${percentage}%` : null;
  const progressBarWidth = percentage > 100 ? 100 : percentage;

  return (
    <div>
      {title}
      <div className="sc-progress">
        <div style={{ width: `${progressBarWidth}%` }} className={getBarClass()}>
          {innerLabel}
        </div>
      </div>
    </div>
  );
}

Progressbar.propTypes = {
  /**
   * title options for title
   */
  titleSettings: PropTypes.shape({
    outside: PropTypes.bool,
    inside: PropTypes.bool
  }),

  /**
   * {Object} color Options for progressbar color
   * @property {Boolean} dynamic Has progressbar a dynamic color range
   * @property {Object} range Settings for dynamic color change
   * @property {Number} min Minimum percentage value needed to apply a class
   * @property {Number} max Maximum percentage needed to apply a class
   * @property {String} className Class name that will be applied based on min and max rules
   */
  color: PropTypes.shape({
    dynamic: PropTypes.bool,
    range: PropTypes.arrayOf(
      PropTypes.shape({
        min: PropTypes.number,
        max: PropTypes.number,
        className: PropTypes.string
      })
    )
  }),

  /**
   * Classes to inherit
   */
  className: PropTypes.string,

  /**
   * Percentage options for title
   */
  percentage: numberInRange(0, 100)
};

Progressbar.defaultProps = {
  className: '',
  titleSettings: {
    inside: false,
    outside: false
  },
  color: {
    dynamic: false
  },
  percentage: 0
};

export default Progressbar;
