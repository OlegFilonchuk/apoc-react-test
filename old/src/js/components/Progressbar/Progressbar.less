.progress-inner-text {
  text-align: center;
  font-size: 12px;
  color: #ffffff;
}
