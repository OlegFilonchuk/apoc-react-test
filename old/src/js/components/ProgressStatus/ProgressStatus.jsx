import React from 'react';
import PropTypes from 'prop-types';

import numberInRange from '../../utils/validation/numberInRange';
import Progressbar from '../Progressbar/Progressbar';
import { STAGE_STATUS } from './ProgressStatusStage';

import './ProgressStatus.less';

function ProgressStatus(props) {
  const { className, children, progressLabel, progress } = props;

  const stageStep = 100 / children.length;

  const baseClass = 'progress-status';
  const progressStatusClasses = `${baseClass} ${className}`;
  const status = index => {
    if (stageStep * (index + 1) <= progress !== stageStep * index <= progress) {
      return STAGE_STATUS.INPROGRESS;
    } else if (stageStep * (index + 1) <= progress) {
      return STAGE_STATUS.FINISHED;
    }

    return STAGE_STATUS.WAITING;
  };

  const stages = React.Children.map(children, (child, index) =>
    React.cloneElement(child, {
      status: status(index)
    })
  );

  return (
    <div className={progressStatusClasses}>
      <div className="sc-row sc-row-no-gutter">
        <div className="sc-col-sm-8 sc-col-sm-3-offset">
          <ul className={`${baseClass}__stages-container`}>{stages}</ul>
        </div>
      </div>
      <div className="sc-row sc-row-no-gutter">
        <div className={`sc-col-sm-3 ${baseClass}__label`}>{progressLabel}</div>
        <div className="sc-col-sm-8">
          <Progressbar percentage={progress} />
        </div>
        <div className={`sc-col-sm-1 ${baseClass}__progress`}>{progress}%</div>
      </div>
    </div>
  );
}

ProgressStatus.propTypes = {
  /**
   * List of custom classes added to component
   */
  className: PropTypes.string,
  /**
   * The label of the progress ex: filename beeing processed
   */
  progressLabel: PropTypes.string.isRequired,
  /**
   * The progress so far in percents
   */
  progress: numberInRange(0, 100),
  /**
   * `ProgressStatusStage`s child nodes
   */
  children: PropTypes.node.isRequired
};

ProgressStatus.defaultProps = {
  progress: 0,
  className: ''
};

export default ProgressStatus;
