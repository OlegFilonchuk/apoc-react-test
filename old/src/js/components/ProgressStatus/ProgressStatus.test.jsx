import React from 'react';
import { mount } from 'enzyme';

import ProgressStatus from './ProgressStatus';
import ProgressStatusStage, { STAGE_STATUS } from './ProgressStatusStage';
import Progressbar from '../Progressbar/Progressbar';

describe('<ProgressStatus />', () => {
  it('renders ProgressStatus component', () => {
    // arrange
    const progress = 30;

    // act
    const component = mount(
      <ProgressStatus progress={progress} progressLabel="some file.txt">
        <ProgressStatusStage title="Importing File" />
        <ProgressStatusStage title="Converting To Scenario" />
        <ProgressStatusStage title="Reloading the view" />
      </ProgressStatus>
    );

    // assert
    expect(component).toMatchSnapshot();
  });

  it('renders Progressbar inside ProgressStatus with correct percentage', () => {
    // arrange
    const progress = 30;

    // act
    const component = mount(
      <ProgressStatus progress={progress} progressLabel="some file.txt">
        <ProgressStatusStage title="Importing File" />
        <ProgressStatusStage title="Converting To Scenario" />
        <ProgressStatusStage title="Reloading the view" />
      </ProgressStatus>
    );

    // assert
    expect(component.find(Progressbar).props().percentage).toEqual(30);
  });

  it('renders ProgressStatusStage right number of child components', () => {
    // arrange
    const progress = 30;

    // act
    const component = mount(
      <ProgressStatus progress={progress} progressLabel="some file.txt">
        <ProgressStatusStage title="Importing File" />
        <ProgressStatusStage title="Converting To Scenario" />
        <ProgressStatusStage title="Reloading the view" />
      </ProgressStatus>
    );

    // assert
    expect(component.find(ProgressStatusStage).length).toEqual(3);
  });

  it('renders ProgressStatusStage right number of child components #2', () => {
    // arrange
    const progress = 30;

    // act
    const component = mount(
      <ProgressStatus progress={progress} progressLabel="some file.txt">
        <ProgressStatusStage title="Importing File" />
        <ProgressStatusStage title="Converting To Scenario" />
      </ProgressStatus>
    );

    // assert
    expect(component.find(ProgressStatusStage).length).toEqual(2);
  });

  it('renders ProgressStatusStage child components in correct status value', () => {
    // arrange
    const progress = 60;

    // act
    const component = mount(
      <ProgressStatus progress={progress} progressLabel="some file.txt">
        <ProgressStatusStage title="Importing File" />
        <ProgressStatusStage title="Converting To Scenario" />
        <ProgressStatusStage title="Reloading the view" />
      </ProgressStatus>
    );

    // assert
    expect(
      component
        .find(ProgressStatusStage)
        .at(0)
        .props().status
    ).toEqual(STAGE_STATUS.FINISHED);
    expect(
      component
        .find(ProgressStatusStage)
        .at(1)
        .props().status
    ).toEqual(STAGE_STATUS.INPROGRESS);
    expect(
      component
        .find(ProgressStatusStage)
        .at(2)
        .props().status
    ).toEqual(STAGE_STATUS.WAITING);
  });
});
