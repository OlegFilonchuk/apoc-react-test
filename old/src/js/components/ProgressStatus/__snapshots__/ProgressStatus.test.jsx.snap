// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressStatus /> renders ProgressStatus component 1`] = `
<ProgressStatus
  className=""
  progress={30}
  progressLabel="some file.txt"
>
  <div
    className="progress-status "
  >
    <div
      className="sc-row sc-row-no-gutter"
    >
      <div
        className="sc-col-sm-8 sc-col-sm-3-offset"
      >
        <ul
          className="progress-status__stages-container"
        >
          <ProgressStatusStage
            className=""
            key=".0"
            status="inprogress"
            title="Importing File"
          >
            <li
              className="progress-status-stage__wrapper"
            >
              <div
                className="progress-status-stage__container progress-status-stage__container--inprogress"
              >
                <span
                  className="progress-status-stage__container-title"
                >
                  Importing File
                </span>
                <div
                  className="progress-status-stage__progress-animation"
                />
              </div>
            </li>
          </ProgressStatusStage>
          <ProgressStatusStage
            className=""
            key=".1"
            status="waiting"
            title="Converting To Scenario"
          >
            <li
              className="progress-status-stage__wrapper"
            >
              <div
                className="progress-status-stage__container"
              >
                <span
                  className="progress-status-stage__container-title"
                >
                  Converting To Scenario
                </span>
              </div>
            </li>
          </ProgressStatusStage>
          <ProgressStatusStage
            className=""
            key=".2"
            status="waiting"
            title="Reloading the view"
          >
            <li
              className="progress-status-stage__wrapper"
            >
              <div
                className="progress-status-stage__container"
              >
                <span
                  className="progress-status-stage__container-title"
                >
                  Reloading the view
                </span>
              </div>
            </li>
          </ProgressStatusStage>
        </ul>
      </div>
    </div>
    <div
      className="sc-row sc-row-no-gutter"
    >
      <div
        className="sc-col-sm-3 progress-status__label"
      >
        some file.txt
      </div>
      <div
        className="sc-col-sm-8"
      >
        <Progressbar
          className=""
          color={
            Object {
              "dynamic": false,
            }
          }
          percentage={30}
          titleSettings={
            Object {
              "inside": false,
              "outside": false,
            }
          }
        >
          <div>
            <div
              className="sc-progress"
            >
              <div
                className="sc-progress-bar"
                style={
                  Object {
                    "width": "30%",
                  }
                }
              />
            </div>
          </div>
        </Progressbar>
      </div>
      <div
        className="sc-col-sm-1 progress-status__progress"
      >
        30
        %
      </div>
    </div>
  </div>
</ProgressStatus>
`;
